README.md
# A Bash Prompt

A Bash prompt written by pure Bash script. Make Bash great again!

All prompt sections are configurable, extendable and easy to use.

## Preview

### Responsive prompt (Auto fit to window width).

![responsive-prompt.png](https://media.githubusercontent.com/media/adoyle-h/_imgs/master/github/a-bash-prompt/responsive-prompt.png)

### Show last command exit status

![exit-status.png](https://media.githubusercontent.com/media/adoyle-h/_imgs/master/github/a-bash-prompt/exit-status.png)

### Highlight Backgound jobs:

![jobs-labels.png](https://media.githubusercontent.com/media/adoyle-h/_imgs/master/github/a-bash-prompt/jobs-labels.png)

### Bubble Style and Block Style

![bubble-and-block-styles.png](https://media.githubusercontent.com/media/adoyle-h/_imgs/master/github/a-bash-prompt/bubble-and-block-styles.png)

### Quick preview in container

You can easily preview it in container.

```sh
# build docker image (default bash 4.4)
./build
# run and enter container (default bash 4.4)
./run

# or build with specific bash version
./build 5.0
./run 5.0
```

## Prerequisites

### Supported Platform

| Supported | Platform | Version | Main Reasons                       |
|:---------:|:---------|:--------|:-----------------------------------|
|     ✅    | MacOS    | *       | -                                  |
|     ✅    | Linux    | *       | -                                  |
|     ✅    | BSD      | *       | -                                  |
|     🚫    | Windows  | -       | Never and ever supported.          |

### Supported Shells

| Supported | Shell | Version          | Main Reasons                                |
|:---------:|:------|:-----------------|:--------------------------------------------|
|     ❔    | Zsh   | v5 and higher    | Not tested yet.                             |
|     ✅    | Bash  | v5 and higher    | -                                           |
|     ✅    | Bash  | v4.2 and higher  | -                                           |
|     🚫    | Bash  | v3.x, v4.0, v4.1 | Associative array not supported  until v4.0 |

## Versions

See [releases][].
The versions follows the rules of [SemVer 2.0.0](http://semver.org/).

## Install

```sh
git clone https://github.com/adoyle-h/a-bash-prompt.git
echo "source $PWD/a-bash-prompt/a.prompt.bash" > ~/.bahsrc
```

## Default Options

You can set these variables before and after `source a.prompt.bash` to override defaults options.

You can modify `PROMPT_NO_COLOR`, `PROMPT_COLOR_*` and `PROMPT_STYLE_*` variables to change the display in runtime.

Available colors: RED GREEN YELLOW BLUE PURPLE CYAN WHITE BLACK GREY
Available styles: bubble block square none
(Case-sensitive)

Belows are default options.

### Others

```sh
PROMPT_NO_COLOR=0               # If set 1, no color printed
PROMPT_NO_MODIFY_LSCOLORS=0
PROMPT_PS1=''                   # If `PROMPT_PS1` is not empty, the PS1 in framework will be override.
```

### Layout

You can rearrange the layout.

```sh
PROMPT_LAYOUT_RIGHT=( exit_status jobs python_virtualenv time )
PROMPT_LAYOUT_LEFT=( left_icon user hostname cwd )
PROMPT_LAYOUT_MAIN=( indicator git reset_text )
```

### Time Clock

```sh
PROMPT_ENABLE_TIME=1
PROMPT_STYLE_TIME=block
PROMPT_COLOR_TIME=YELLOW
PROMPT_FORMAT_TIME='T%s'        # You can set ' %s'
PROMPT_PS1_TIME_DATE_FORMAT='+%H:%M:%S' # date +%H:%M:%S
```

### Front and Backgound Jobs

```sh
PROMPT_ENABLE_JOB=1
PROMPT_STYLE_JOB=block
PROMPT_COLOR_JOB=CYAN
PROMPT_FORMAT_JOB='Jobs %s'
```

### Process Exit Status Code

```sh
PROMPT_ENABLE_EXIT_STATUS=1
PROMPT_STYLE_EXIT_STATUS=block
PROMPT_COLOR_EXIT_STATUS=RED
PROMPT_FORMAT_EXIT_STATUS='😱 %s'
```

### Current Work Directory

```sh
PROMPT_ENABLE_CWD=1
PROMPT_STYLE_CWD=block
PROMPT_COLOR_CWD=GREEN
PROMPT_FORMAT_CWD=' %s '
```

### Current User

```sh
PROMPT_ENABLE_USER=0
PROMPT_STYLE_USER=square
PROMPT_COLOR_USER=CYAN
PROMPT_FORMAT_USER='%s'
```

### Hostname

```sh
PROMPT_ENABLE_HOSTNAME=0
PROMPT_STYLE_HOSTNAME=square
PROMPT_COLOR_HOSTNAME=PURPLE
PROMPT_FORMAT_HOSTNAME='%s'
```

### Git branch and status

```sh
PROMPT_ENABLE_GIT=1
PROMPT_COLOR_GIT=BLUE
PROMPT_STYLE_GIT=none
PROMPT_FORMAT_GIT='%b'
# See https://github.com/git/git/blob/master/contrib/completion/git-prompt.sh
GIT_PS1_SHOWDIRTYSTATE=1
GIT_PS1_SHOWSTASHSTATE=1
GIT_PS1_SHOWUNTRACKEDFILES=1
GIT_PS1_SHOWCOLORHINTS=true
GIT_PS1_SHOWUPSTREAM=git
GIT_PS1_DESCRIBE_STYLE=branch
GIT_PS1_STATESEPARATOR=' '
```

### Python Virtual Environment

```sh
PROMPT_ENABLE_PYTHON_VENV=1
PROMPT_STYLE_PYTHON_VENV=block
PROMPT_COLOR_PYTHON_VENV=PURPLE
PROMPT_FORMAT_PYTHON_VENV='%s'
PROMPT_PYTHON_VIRTUALENV_LEFT='venv:'
```

### Left Icon

```sh
PROMPT_ENABLE_LEFT_ICON=1
PROMPT_STYLE_LEFT_ICON=none
PROMPT_COLOR_LEFT_ICON=GREEN
PROMPT_FORMAT_LEFT_ICON='%s '
PROMPT_PS1_LEFT_ICON='⧉ '
```

### Indicator

```sh
PROMPT_ENABLE_INDICATOR=1
PROMPT_STYLE_INDICATOR=none
PROMPT_COLOR_INDICATOR=GREEN
PROMPT_FORMAT_INDICATOR='%s '
PROMPT_PS1_INDICATOR='𝕬'
```

### History Append

call `history -a` each time.

```sh
PROMPT_ENABLE_HISTORY_APPEND=0
```

## Advanced Usage

### Write section to extend prompt

```sh
__ps1_section_hello() {
  [[ ${PROMPT_ENABLE_HELLO:-1} == 0 ]] && return
  local PROMPT_STYLE_HELLO=${PROMPT_STYLE_HELLO:-block}
  local PROMPT_COLOR_HELLO=${PROMPT_COLOR_HELLO:-GREEN}
  local PROMPT_FORMAT_HELLO=${PROMPT_FORMAT_HELLO:-'%s'}
  __ps1_print_section HELLO "hello world"
}

PROMPT_LAYOUT_RIGHT+=( hello )
```

Then you will get these.

![write-your-section.png](https://media.githubusercontent.com/media/adoyle-h/_imgs/master/github/a-bash-prompt/write-your-section.png)

### Git-prompt is slow

If you feel slow in git directory. Disable these options will make it faster.

```sh
GIT_PS1_SHOWDIRTYSTATE=
GIT_PS1_SHOWSTASHSTATE=
GIT_PS1_SHOWUNTRACKEDFILES=
```

See https://github.com/git/git/blob/master/contrib/completion/git-prompt.sh

You can also disable git-prompt completely by `unset __git_ps1`.

### Override PS1

If `PROMPT_PS1` set, the PS1 in framework will be override.

```sh
PROMPT_PS1='\u@\h:\w\$ '
```

### Disable prompt color

Set `PROMPT_NO_COLOR=1`.

![no-color.png](https://media.githubusercontent.com/media/adoyle-h/_imgs/master/github/a-bash-prompt/no-color.png)

### Do not modify LS_COLORS variable

a-bash-prompt will modify `LSCOLORS` and `LS_COLORS` by default.

Set `PROMPT_NO_MODIFY_LSCOLORS=1` to disable it.

### Refresh bash_history

bash_history file will not be refreshed before exit shell by default.

If you use tmux, it may cause some troubles.
Set `PROMPT_ENABLE_HISTORY_APPEND=1` to refresh bash_history each command.

## Suggestion, Bug Reporting, Contributing

**Before opening new Issue/Discussion/PR and posting any comments**, please read [Contributing Guidelines](https://gcg.adoyle.me/CONTRIBUTING).

## Copyright and License

Copyright 2019-2023 ADoyle (adoyle.h@gmail.com). Some Rights Reserved.
The project is licensed under the **BSD 3-clause License**.

See the [LICENSE][] file for the specific language governing permissions and limitations under the License.

See the [NOTICE][] file distributed with this work for additional information regarding copyright ownership.

## Other Projects

- [shell-general-colors](https://github.com/adoyle-h/shell-general-colors): To generate [colors.bash](./colors.bash).
- [one.share](https://github.com/one-bash/one.share)
- [Other shell projects](https://github.com/adoyle-h?tab=repositories&q=&type=source&language=shell&sort=stargazers) created by me.


<!-- links -->

[LICENSE]: ./LICENSE
[NOTICE]: ./NOTICE
[releases]: https://github.com/adoyle-h/a-bash-prompt/releases


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is inspired by [Keep a Changelog 1.0.0](https://keepachangelog.com/en/1.0.0/).

The versions follow the rules of [Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

[Full Changes](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/compare/v2.0.0...HEAD)


<a name="v2.0.0"></a>
## v2.0.0 (2023-06-08 17:12:17 +08:00)

[Full Changes](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/compare/v1.3.1...v2.0.0)

### Breaking Changes

Have 1 breaking changes. Check below logs with ⚠️ .

### New Features

- support PROMPT_PS1_TIME_DATE_FORMAT ([a50bb66](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/a50bb6616eb2c8d0b83c460a09cb05e8520822b6))
  > Defaults to '+%H:%M:%S'. It is used for `date +%H:%M:%S`.
- user can change the section layout ([8d258e3](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/8d258e3c93336e4c26ceb7636b88697ec01ec065))
  > Default Layout:
  > - PROMPT_LAYOUT_RIGHT=( exit_status jobs python_virtualenv time )
  > - PROMPT_LAYOUT_LEFT=( left_icon user hostname cwd )
  > - PROMPT_LAYOUT_MAIN=( indicator git reset_text )
- ⚠️  add INDICATOR ([14df408](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/14df408312023bf84af188e31e59cd87d36162b0))
  > - support PROMPT_ENABLE_INDICATOR. Is is enabled by default.
  > 
  > Breaking Change:
  > 
  > - using PROMPT_PS1_INDICATOR instead of PROMPT_PS1_SIGNATURE
  > - using PROMPT_COLOR_INDICATOR instead of PROMPT_COLOR_SIGNATURE
- add USER ([a0bba0b](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/a0bba0b6e403b47fd08afc29ce3d4c14c8480bfc))
  > - Enable it via `PROMPT_ENABLE_USER=1`. It is disabled by default.
- add HOSTNAME and PROMPT_COLOR_BG ([29cec72](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/29cec72bc990c47813436f35d4bcb5a8b92cfe2b))
  > - Set `PROMPT_ENABLE_HOSTNAME=1` to enable it. default to 0.
  > - `PROMPT_STYLE_HOSTNAME=square`. support square or bubble or block
  > - `PROMPT_COLOR_HOSTNAME=PURPLE`
  > - User can change the background color of prompt via `PROMPT_COLOR_BG=BLACK`

### Document Changes

- better document ([c2cefe6](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/c2cefe6e8a2e2ed008366aeb029a3cbd03e9b574))

<a name="v1.3.1"></a>
## v1.3.1 (2023-02-10 14:59:56 +08:00)

[Full Changes](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/compare/v1.3.0...v1.3.1)

### Document Changes

- renew license date && update contributing ([46c7037](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/46c703789d08a2bea42bac493476794484defc84))
- add docs and Github Issue Templates ([cac9c46](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/cac9c46586b6a16223aa2db9f3b67155698e6069))

<a name="v1.3.0"></a>
## v1.3.0 (2022-11-22 20:23:30 +08:00)

[Full Changes](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/compare/v1.2.1...v1.3.0)

### New Features

- support user defined FORMAT ([1fb1c05](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/1fb1c0524d0507145b9662c287a250ba1b3775ea))
  > add PROMPT_FORMAT_CWD PROMPT_FORMAT_TIME PROMPT_FORMAT_EXIT_STATUS PROMPT_FORMAT_JOB

<a name="v1.2.1"></a>
## v1.2.1 (2022-11-10 16:49:39 +08:00)

[Full Changes](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/compare/v1.2.0...v1.2.1)

### Bug Fixes

- use RESET_BG instead of BG_BLACK ([474ca8a](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/474ca8a277922c254726a01c53fa1f4f59a39cdf))

<a name="v1.2.0"></a>
## v1.2.0 (2022-10-27 19:55:50 +08:00)

[Full Changes](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/compare/v1.1.0...v1.2.0)

### New Features

- support PROMPT_COLOR_* variables ([bb08887](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/bb088870665853daef52fca90456ed20a100a6bb))

<a name="v1.1.0"></a>
## v1.1.0 (2022-10-18 11:37:40 +08:00)

[Full Changes](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/compare/v1.0.1...v1.1.0)

### New Features

- new prompt style "bubble" ([f346948](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/f346948df03ee6c7c334374dc0ade95bdae62c6d))
  > You can set variables PROMPT_STYLE_CWD, PROMPT_STYLE_TIME, PROMPT_STYLE_EXIT_STATUS, PROMPT_STYLE_JOB.
  > Old style named "block".
- add python virtualenv section ([2181415](https://github.com/https://github.com/adoyle-h/a-bash-prompt.git/commit/2181415fd18bf057342077ae354592db696bcdaa))

### Bug Fixes


a.prompt.bash
#shellcheck disable=SC2155,SC2148

# If not running interactively, don't do anything
[[ -z "$PS1" ]] && return

# Set 0 to disable, set 1 to enable PROMPT_ options
PROMPT_COLOR_BG=${PROMPT_COLOR_BG:-BLACK}
PROMPT_NO_COLOR=${PROMPT_NO_COLOR:-0}
PROMPT_NO_MODIFY_LSCOLORS=${PROMPT_NO_MODIFY_LSCOLORS:-0}

if ! declare -p PROMPT_LAYOUT_RIGHT &>/dev/null ; then
  PROMPT_LAYOUT_RIGHT=( exit_status jobs python_virtualenv time )
fi
if ! declare -p PROMPT_LAYOUT_LEFT &>/dev/null ; then
  PROMPT_LAYOUT_LEFT=( left_icon user hostname cwd )
fi
if ! declare -p PROMPT_LAYOUT_MAIN &>/dev/null ; then
  PROMPT_LAYOUT_MAIN=( indicator git reset_text )
fi

if [[ "$PROMPT_NO_COLOR" != 1 ]]; then
  export CLICOLOR=1

  if [[ $PROMPT_NO_MODIFY_LSCOLORS != 1 ]]; then
    # BSD LSCOLORS
    export LSCOLORS=exgxFxDxcxdgDEHbHDacad
    # Linux LS_COLORS
    export LS_COLORS='di=34:ln=36:so=1;35:pi=1;33:ex=32:bd=33;46:cd=1;33;1;44:su=1;37;41:sg=1;37;1;43:tw=30;42:ow=30;43'
  fi
fi

source "$(dirname "${BASH_SOURCE[0]}")"/colors.bash

__prompt_os_kernel=$(uname -s)

#######################################################################
#                          Helper Fcuntions                           #
#######################################################################

if [[ $__prompt_os_kernel == Darwin ]]; then
  # grep not support unicode well in MacOS. So use perl instead of.
  __prompt_fix_middle_length() {
    local raw=$1
    # Match Chinese and Emoji characters
    local l=$(perl -CS -pe 's/[\x{4E00}-\x{9FA5}\x{1F601}-\x{1F64F}]+//g' <<<"$raw")
    printf '%s\n' "$((${#raw} - ${#l}))"
  }
else
  __prompt_fix_middle_length() {
    grep -oE $'[\u4e00-\u9fa5😱]' <<<"$plain" | wc -l | tr -d ' ' || true
  }
fi

__prompt_debug() {
  echo "$@" >>~/prompt_debug
}

# http://jafrog.com/2013/11/23/colors-in-terminal.html
__prompt_trim_str_color() {
  local ecs=$'\e'
  if [[ $__prompt_os_kernel == Darwin ]]; then
    # awful BSD sed
    sed -E "s,\\${ecs}[[0-9]*(;[0-9]+)*m,,g" <<<"$1"
  else
    sed -E "s,${ecs}[[0-9]*(;[0-9]+)*m,,g" <<<"$1"
  fi
}

# Wrapping in \[ \] is recommended by the Bash man page.
# This helps Bash ignore non-printable characters so that it correctly calculates the size of the prompt.
__prompt_wrap_color() {
  local ecs=$'\e'
  local start='\\['
  local end='\\]'
  if [[ $__prompt_os_kernel == Darwin ]]; then
    sed -E "s,\\${ecs}[[0-9]*(;[0-9]+)*m,${start}&${end},g" <<<"$1"
  else
    sed -E "s,${ecs}[[0-9]*(;[0-9]+)*m,${start}\\0${end},g" <<<"$1"
  fi
}

__prompt_check_precmd_conflict() {
  local f
  for f in "${precmd_functions[@]}"; do
    if [[ "${f}" == "${1}" ]]; then
      return 0
    fi
  done
  return 1
}

# Copy from bash-it. https://github.com/Bash-it/bash-it/blob/master/themes/base.theme.bash
__prompt_append() {
  local prompt_re

  if [[ -n "${bash_preexec_imported:-}" ]]; then
    # We are using bash-preexec
    if ! __prompt_check_precmd_conflict "${1}"; then
      precmd_functions+=("${1}")
    fi
  else
    # Set OS dependent exact match regular expression
    if [[ ${OSTYPE} == darwin* ]]; then
      # macOS
      prompt_re="[[:<:]]${1}[[:>:]]"
    else
      # Linux, FreeBSD, etc.
      prompt_re="\\<${1}\\>"
    fi

    if [[ ${PROMPT_COMMAND:-} =~ ${prompt_re} ]]; then
      return
    elif [[ -z ${PROMPT_COMMAND:-} ]]; then
      PROMPT_COMMAND="${1}"
    else
      PROMPT_COMMAND="${1};${PROMPT_COMMAND}"
    fi
  fi
}

#######################################################################
#                         Section Definitions                         #
#######################################################################

# PROMPT_STYLE_SECTION=square|bubble|block|none
# PROMPT_COLOR_SECTION=BLACK|WHITE|YELLOW|CYAN|BLUE|GREEN|GREY|RED|PURPLE
# PROMPT_FORMAT_SECTION='%s'
__ps1_print_section() {
  local -n style=PROMPT_STYLE_$1
  local -n color=PROMPT_COLOR_$1
  local -n format=PROMPT_FORMAT_$1
  local value=$2

  [[ -z $value ]] && return

  local fg="${__prompt_colors[${color}]}"

  if [[ $style == bubble ]]; then
    printf "%b%b$format%b" \
      "${fg}" \
      "${__prompt_colors[BG_${color}]}${__prompt_colors[$PROMPT_COLOR_BG]}" \
      "$value" \
      "${__prompt_colors[RESET_BG]}${fg}"
  elif [[ $style == square ]]; then
    printf "%b $format %b" \
      "${__prompt_colors[BG_${color}]}${__prompt_colors[$PROMPT_COLOR_BG]}" \
      "$value" \
      "${__prompt_colors[RESET_BG]}${fg}"
  elif [[ $style == block ]]; then
    printf "%b[$format]" "${fg}" "$value"
  else
    printf "%b$format" "${fg}" "$value"
  fi
}

__ps1_section_exit_status() {
  [[ ${PROMPT_ENABLE_JOB:-1} == 0 ]] && return

  local exit_status=$__ps1_last_exit_status

  if [[ $exit_status != 0 ]]; then
    local PROMPT_STYLE_EXIT_STATUS=${PROMPT_STYLE_EXIT_STATUS:-block}
    local PROMPT_COLOR_EXIT_STATUS=${PROMPT_COLOR_EXIT_STATUS:-RED}
    local PROMPT_FORMAT_EXIT_STATUS=${PROMPT_FORMAT_EXIT_STATUS:-'😱 %s'}
    __ps1_print_section EXIT_STATUS "$exit_status"
  fi
}

__ps1_section_jobs() {
  [[ ${PROMPT_ENABLE_JOB:-1} == 0 ]] && return

  local stopped=$(jobs -sp | wc -l | tr -d ' ')
  local running=$(jobs -rp | wc -l | tr -d ' ')

  if ((running > 0)) || ((stopped > 0)); then
    local PROMPT_STYLE_JOB=${PROMPT_STYLE_JOB:-block}
    local PROMPT_COLOR_JOB=${PROMPT_COLOR_JOB:-CYAN}
    local PROMPT_FORMAT_JOB=${PROMPT_FORMAT_JOB:-'Jobs %s'}
    __ps1_print_section JOB "${running:-0}|${stopped:-0}"
  fi
}

__ps1_section_hostname() {
  [[ ${PROMPT_ENABLE_HOSTNAME:-0} == 0 ]] && return
  local PROMPT_STYLE_HOSTNAME=${PROMPT_STYLE_HOSTNAME:-square}
  local PROMPT_COLOR_HOSTNAME=${PROMPT_COLOR_HOSTNAME:-PURPLE}
  local PROMPT_FORMAT_HOSTNAME=${PROMPT_FORMAT_HOSTNAME:-'%s'}
  __ps1_print_section HOSTNAME "$HOSTNAME"
}

__ps1_section_user() {
  [[ ${PROMPT_ENABLE_USER:-0} == 0 ]] && return
  local PROMPT_STYLE_USER=${PROMPT_STYLE_USER:-square}
  local PROMPT_COLOR_USER=${PROMPT_COLOR_USER:-CYAN}
  local PROMPT_FORMAT_USER=${PROMPT_FORMAT_USER:-'%s'}
  __ps1_print_section USER "$USER"
}

__ps1_section_time() {
  [[ ${PROMPT_ENABLE_TIME:-1} == 0 ]] && return
  local PROMPT_STYLE_TIME=${PROMPT_STYLE_TIME:-block}
  local PROMPT_COLOR_TIME=${PROMPT_COLOR_TIME:-YELLOW}
  local PROMPT_FORMAT_TIME=${PROMPT_FORMAT_TIME:-'T%s'}
  __ps1_print_section TIME "$(date "${PROMPT_PS1_TIME_DATE_FORMAT:-+%H:%M:%S}")"
}

__ps1_section_left_icon() {
  [[ ${PROMPT_ENABLE_LEFT_ICON:-1} == 0 ]] && return
  local PROMPT_STYLE_LEFT_ICON=${PROMPT_STYLE_LEFT_ICON:-none}
  local PROMPT_COLOR_LEFT_ICON=${PROMPT_COLOR_LEFT_ICON:-GREEN}
  local PROMPT_FORMAT_LEFT_ICON=${PROMPT_FORMAT_LEFT_ICON:-'%s '}
  __ps1_print_section LEFT_ICON "${PROMPT_PS1_LEFT_ICON:-⧉}"
}

__ps1_section_cwd() {
  [[ ${PROMPT_ENABLE_CWD:-1} == 0 ]] && return
  local PROMPT_STYLE_CWD=${PROMPT_STYLE_CWD:-block}
  local PROMPT_COLOR_CWD=${PROMPT_COLOR_CWD:-GREEN}
  local PROMPT_FORMAT_CWD=${PROMPT_FORMAT_CWD:-' %s '}
  __ps1_print_section CWD "$PWD"
}

__ps1_section_git() {
  [[ ${PROMPT_ENABLE_GIT:-1} == 0 ]] && return

  if command -v __git_ps1 &>/dev/null; then
    # Set empty string to disable, set non-empty string to enable GIT_ options
    local GIT_PS1_SHOWDIRTYSTATE=${GIT_PS1_SHOWDIRTYSTATE:-1}
    local GIT_PS1_SHOWSTASHSTATE=${GIT_PS1_SHOWSTASHSTATE:-1}
    local GIT_PS1_SHOWUNTRACKEDFILES=${GIT_PS1_SHOWUNTRACKEDFILES:-1}
    local GIT_PS1_SHOWCOLORHINTS=${GIT_PS1_SHOWCOLORHINTS:-true}
    local GIT_PS1_SHOWUPSTREAM=${GIT_PS1_SHOWUPSTREAM:-git}
    local GIT_PS1_DESCRIBE_STYLE=${GIT_PS1_DESCRIBE_STYLE:-branch}
    local GIT_PS1_STATESEPARATOR=${GIT_PS1_STATESEPARATOR:-' '}

    local PROMPT_COLOR_GIT=${PROMPT_COLOR_GIT:-BLUE}
    local PROMPT_STYLE_GIT=${PROMPT_STYLE_GIT:-none}
    local PROMPT_FORMAT_GIT=${PROMPT_FORMAT_GIT:-' %b'}
    __ps1_print_section GIT "$(__git_ps1 "(%s)")"
  fi
}

__ps1_section_indicator() {
  [[ ${PROMPT_ENABLE_INDICATOR:-1} == 0 ]] && return
  local PROMPT_STYLE_INDICATOR=${PROMPT_STYLE_INDICATOR:-none}
  local PROMPT_COLOR_INDICATOR=${PROMPT_COLOR_INDICATOR:-GREEN}
  local PROMPT_FORMAT_INDICATOR=${PROMPT_FORMAT_INDICATOR:-'%s'}
  __ps1_print_section INDICATOR "${PROMPT_PS1_INDICATOR:-𝕬}"
}

__ps1_section_fill_middle_spaces() {
  # For debug
  # local left=$1
  # local right=$2
  # local left_plain=$(__prompt_trim_str_color "$left")
  # local right_plain=$(__prompt_trim_str_color "$right")
  # local -i leftDbCharLen=$(grep -oE $'[\u4e00-\u9fa5]' <<< "$left_plain" | wc -l | tr -d ' ' || true)
  # local -i rightDbCharLen=$(grep -oE $'[\u4e00-\u9fa5]' <<< "$right_plain" | wc -l | tr -d ' ' || true)
  # local -i COLS=$(( COLUMNS - ${#left_plain} - ${#right_plain} - leftDbCharLen - rightDbCharLen ))

  # __prompt_debug "COLUMNS=$COLUMNS"
  # __prompt_debug "left.len=${#left} left_plain.len=${#left_plain}"
  # __prompt_debug "right.len=${#right} right_plain.len=${#right_plain}"
  # __prompt_debug "leftDbCharLen=$leftDbCharLen rightDbCharLen=$rightDbCharLen"
  # __prompt_debug "COLS=$COLS"
  # __prompt_debug -e "left=$left"
  # __prompt_debug -e "right=$right"
  # __prompt_debug -e "left_plain=$left_plain"
  # __prompt_debug -e "right_plain=$right_plain"

  local plain=$(__prompt_trim_str_color "$1$2")
  local fixLen=$(__prompt_fix_middle_length "$plain")
  local -i COLS=$((COLUMNS - ${#plain} - fixLen))

  if ((COLS < 1)); then
    printf '\n %b' "${__prompt_colors[GREEN]}➥"
  else
    local LINE=''
    local CHAR='—'
    while ((${#LINE} < COLS)); do
      LINE="$LINE$CHAR"
    done

    printf '%b' "${__prompt_colors[GREY]}${LINE}"
  fi
}

__ps1_section_reset_text() {
  printf '%b' "${__prompt_colors[RESET_ALL]}"
}

__ps1_section_python_virtualenv() {
  [[ ${PROMPT_ENABLE_PYTHON_VENV:-1} == 0 ]] && return

  if [[ -n "$VIRTUAL_ENV" ]]; then
    local PROMPT_STYLE_PYTHON_VENV=${PROMPT_STYLE_PYTHON_VENV:-block}
    local PROMPT_COLOR_PYTHON_VENV=${PROMPT_COLOR_PYTHON_VENV:-PURPLE}
    local PROMPT_FORMAT_PYTHON_VENV=${PROMPT_FORMAT_PYTHON_VENV:-'%s'}
    __ps1_print_section PYTHON_VENV "${PROMPT_PYTHON_VIRTUALENV_LEFT:-venv:} ${VIRTUAL_ENV##*/}"
  fi
}

#######################################################################
#                         Command Definitions                         #
#######################################################################

__ps1_command_append_history() {
  [[ ${PROMPT_ENABLE_HISTORY_APPEND:-0} == 0 ]] && return
  history -a
}

#######################################################################
#                            Presentation                             #
#######################################################################

__ps1_right() {
  local section
  for section in "${PROMPT_LAYOUT_RIGHT[@]}" ; do
    "__ps1_section_$section"
  done
}

__ps1_left() {
  local section
  for section in "${PROMPT_LAYOUT_LEFT[@]}" ; do
    "__ps1_section_$section"
  done
}

__ps1_main() {
  local section
  for section in "${PROMPT_LAYOUT_MAIN[@]}" ; do
    "__ps1_section_$section"
  done
}

__prompt_command() {
  # This line must be first
  local __ps1_last_exit_status=$?

  if [[ -z ${PROMPT_PS1:-} ]]; then
    local right="$(__ps1_right)"
    local left="$(__ps1_left)"
    local middle="$(__ps1_section_fill_middle_spaces "${left}" "${right}")"
    local main="$(__ps1_main)"
    local _PS1="${left:-}${middle:-}${right:-}\\n$(__prompt_wrap_color "$main") "

    if [[ $PROMPT_NO_COLOR == 1 ]]; then
      _PS1=$(__prompt_trim_str_color "$_PS1")
    fi

    PS1=$_PS1
  else
    PS1=$PROMPT_PS1
  fi

  __ps1_command_append_history
}

__prompt_append __prompt_command
